### DISCUSSION TITLE
Error trying to load Microsoft.Bcl.AsyncInterfaces version 1.0.0

### DISCUSSION THREAD

**Discussion by madmoravian** [FortuneN/FineCodeCoverage#112] [General]
I've just recently (the last week or so) been unable to get FCC to complete a run.  It seems to error out because of a lack of Microsoft.Bcl.AsyncInterfaces version 1.0.0  we have 5.0.0 installed in our solution.

Any thoughts?

Here is a snippet from the diagnostics.log file

Resolving assembly.
TpTrace Information: 0 : 43392, 1, 2021/03/11, 12:24:52.297, 12004370685787, vstest.console.dll, AssemblyResolver.OnResolve: Microsoft.Bcl.AsyncInterfaces: Resolved from cache.
TpTrace Verbose: 0 : 43392, 1, 2021/03/11, 12:24:52.297, 12004370686267, vstest.console.dll, CurrentDomain_AssemblyResolve: Resolving assembly 'Microsoft.Bcl.AsyncInterfaces'.
TpTrace Verbose: 0 : 43392, 1, 2021/03/11, 12:24:52.297, 12004370686810, vstest.console.dll, CurrentDomainAssemblyResolve: Failed to resolve assembly 'Microsoft.Bcl.AsyncInterfaces'.
TpTrace Verbose: 0 : 43392, 1, 2021/03/11, 12:24:52.299, 12004370714439, vstest.console.dll, CurrentDomainAssemblyResolve: Failed to resolve assembly 'Microsoft.Bcl.AsyncInterfaces'.
TpTrace Warning: 0 : 43392, 1, 2021/03/11, 12:24:52.319, 12004370910913, vstest.console.dll, TestPluginDiscoverer: Failed to get types from assembly 'Microsoft.Diagnostics.NETCore.Client, Version=0.2.2.41902, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.  Skipping test extension scan for this assembly.  Error: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.
   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework.TestPluginDiscoverer.GetTestExtensionsFromAssembly[TPluginInfo,TExtension](Assembly assembly, Dictionary`2 pluginInfos)
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.

**Comment by tonyhallett** [comment]
Unfortunately this means nothing to me.
What is the output from FCC output window ?
I assume that your tests run normally ?
Did this project work with a previous version of FCC ?
What is Microsoft.Bcl.AsyncInterfaces used for ?
https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces/

> This package is not required starting with .NET Standard 2.1 and .NET Core 3.0.

Can you update ?

**Reply by madmoravian** [reply]
Answer to Question 1: Output from FCC in the output window is:


Fine Code Coverage : ================================== START ==================================

Fine Code Coverage : Run Coverage Tool (GatewayDataServices.IntegrationTests)

Fine Code Coverage : Coverlet Run (GatewayDataServices.IntegrationTests) Arguments 
"C:\Users\wzrubek\Documents\Gateway\gatewaydataservices\GatewayDataServices.IntegrationTests\bin\Debug\netcoreapp3.1\fine-code-coverage\build-output\GatewayDataServices.IntegrationTests.dll"
--format "cobertura"
--exclude-by-file "**/Migrations/*"
--exclude-by-file "**/Models/*"
--exclude-by-file "**/Views/*"
--exclude-by-file "**/ViewModels/*"
--exclude-by-file "**/XMLServices/*"
--exclude-by-attribute "GeneratedCode"
--include-test-assembly
--target "dotnet"
--threshold-type line
--threshold-stat total
--threshold 0
--output "C:\Users\wzrubek\Documents\Gateway\gatewaydataservices\GatewayDataServices.IntegrationTests\bin\Debug\netcoreapp3.1\fine-code-coverage\coverage-tool-output\project.coverage.xml"
--targetargs "test  ""C:\Users\wzrubek\Documents\Gateway\gatewaydataservices\GatewayDataServices.IntegrationTests\bin\Debug\netcoreapp3.1\fine-code-coverage\build-output\GatewayDataServices.IntegrationTests.dll"" --nologo --blame  --results-directory ""C:\Users\wzrubek\Documents\Gateway\gatewaydataservices\GatewayDataServices.IntegrationTests\bin\Debug\netcoreapp3.1\fine-code-coverage\coverage-tool-output"" --diag ""C:\Users\wzrubek\Documents\Gateway\gatewaydataservices\GatewayDataServices.IntegrationTests\bin\Debug\netcoreapp3.1\fine-code-coverage\coverage-tool-output/diagnostics.log""  "

And nothing else ever appears.

Answer to question 2: Yes, our test run normally

Answer to question 3: This project worked with FCC as of last Thursday.  We've not updated FCC recently.  We have started to convert more of our methods over to be async

Asnwer to question 4: probably for dealing with async methods for database intensive applications.

Asnwer to question 5: we are on the latest version of the library

**Comment by tonyhallett** [comment]
The changes introduced recently are including dependencies ( coverlet / open cover and report generator ) as zips and updating to version 3.0.3 of coverlet.

To determine the cause of the issue I suggest.

You supply your own version of the coverlet collector - we can then determine if it is the zip, then if it is the version.

There are 3 visual studio options for this.
CoverletConsoleLocal			Specify true to use your own dotnet tools local install of coverlet console.
CoverletConsoleCustomPath		Specify path to coverlet console exe if you need functionality that the FCC version does not provide.
CoverletConsoleGlobal			Specify true to use your own dotnet tools global install of coverlet console.

Use the first or third option ( simpler option ) and use version 3.0.3.
[coverlet console](https://github.com/coverlet-coverage/coverlet#net-global-tool-guide-suffers-from-possible-known-issue)

If the error still occurs then install version 1.7.2 and try again.

You could also run the tool outside of FCC with the arguments as seen in the FCC output.

The other option is to set the project up for running with coverlet collector.
From the [README](https://github.com/FortuneN/FineCodeCoverage/blob/master/README.md)

> Coverlet has different "drivers".  Fine Code Coverage has in the past only used the coverlet console driver.  This has some issues associated with it.
It is now possible to switch to the Data Collector driver.  This is the better driver but cannot be used for all projects.
For now this is opt in.  In the future Fine Code Coverage will determine the appropriate driver.
Please consult [coverlet docs](https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/VSTestIntegration.md) for version support.
** Note that it is unnecessary to add the nuget coverlet.collector package as FCC internally supplies coverlet.collector 3.0.3**
Fine Code Coverage will use the Data Collector driver under two circumstances :
1) You are testing with runsettings that contains the coverlet collector ( and not disabled)
2) You set the UseDataCollector project property

**Comment by tonyhallett** [comment]
I suspect that the issue is with the installing of the tool ( and then the zip ).

There is a flag when using dotnet tool install 
https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-tool-install

> --framework <FRAMEWORK>
Specifies the target framework to install the tool for. By default, the .NET SDK tries to choose the most appropriate target framework.

**By default, the .NET SDK tries to choose the most appropriate target framework.**

https://github.com/dotnet/sdk/blob/3015692c6c3748b7d13387cbecbb84604b633b77/src/Cli/dotnet/ToolPackage/ToolPackageInstaller.cs#L38

A temp project is created for installing the package

> targetFramework: string.IsNullOrEmpty(targetFramework) ? BundledTargetFramework.GetTargetFrameworkMoniker() :  targetFramework,

and
https://github.com/dotnet/sdk/blob/b91b88aec2684e3d2988df8d838d3aa3c6240a35/src/Cli/dotnet/BundledTargetFramework.cs
```
namespace Microsoft.DotNet.Cli
{
    internal static class BundledTargetFramework
    {
        public static string GetTargetFrameworkMoniker()
        {
            TargetFrameworkAttribute targetFrameworkAttribute = typeof(BundledTargetFramework)
                .GetTypeInfo()
                .Assembly
                .GetCustomAttribute<TargetFrameworkAttribute>();

            return NuGetFramework
                .Parse(targetFrameworkAttribute.FrameworkName)
                .GetShortFolderName();
        }
    }
}
```

On the machine that I installed coverlet.console, prior to zipping, net 5.0 is installed and you can see that in the unzipped path.
\AppData\Local\FineCodeCoverage\coverlet\3.0.3\.store\coverlet.console\3.0.3\coverlet.console\3.0.3\tools\net5.0\any

If you can do as requested then we should know if this is the cause and FCC can return to installing from dotnet tool on the user machine.

**Reply by madmoravian** [reply]
I've installed the coverlet console 3.0.3 and set the flag to use the global coverlet to no effect. 

The tool still never returns any information to us about coverage.  

I've been unable to downgrade the console to 1.7.2

We are on .Net 3.1 and not 5.0

This seems to have started in the middle of last week.  It was working fine up until then.

**Reply by tonyhallett** [reply]
Why can't you install the previous version ?
https://www.nuget.org/packages/coverlet.console/1.7.2

**Reply by madmoravian** [reply]
It errored out attempting to uninstall the old version and installing the new version last week.  Last night it succeeded and I was able to install 1.7.2.

This morning FCC ran to completion.  I have a colleague attempting to duplicate the experiment.

**Reply by madmoravian** [reply]
Rolling back to FCC 1.1.127 also worked for us.

**Reply by tonyhallett** [reply]
The version of coverlet console that FCC installs, 3.0.3, requires .NET 5 to be installed in your machine.  If you had .NET 5 installed and was receiving 'Error trying to load Microsoft.Bcl.AsyncInterfaces version 1.0.0' then this would be a coverlet bug.

**Comment by madmoravian** [comment]
After upgrading to .Net 5.0 we had the same issue.  We updated to the latest version of FCC to no success.  We discovered that the run settings for the Test Explorer were not properly set.  After setting those values, FCC ran flawlessly.  It looks like the error in the diagnostics.log was a red herring.

**Reply by tonyhallett** [reply]
Ok so the problem was not FCC related ? Does it work with coverlet console 3.0.3 ?

**Reply by madmoravian** [reply]
Yes, I've updated to 3.0.3 and I think other folks are using the console that comes with the installation.

**Reply by tonyhallett** [reply]
Excellent. For future reference, the data collector approach, where possible, is the preferred option.
